### ISSUE TITLE
A Single IntVecCounter doesn't increase while others still do

### ISSUE THREAD

**Issue by xMAC94x** [tikv/rust-prometheus#321]
**Describe the bug**
I have prometheus + async-std + rouille in a project.
There is a weird constellation that a IntVecCounter isn't updated any more in async context.
When the webserver runs `registry.gather()` it doesn`t update the value of a specific metric.
I tried to create a minimal reproduction scenario, however i can only reproduce it with a specific 
commit of my main project. steps to reproduce bellow.
Part of my coding is:
```rust
#network/src/participant.rs:171
        let x = self.metrics.frames_in_total.with_label_values(&["dasd", "dd", "ParticipantId"]);
        x.inc(); // <- 1
        let send_outgoing = { send_outgoing.lock().unwrap().clone() };
        let mut messages = HashMap::new();
        let pid_u128: u128 = self.remote_pid.into();
        let pid_string = pid_u128.to_string();
        x.inc(); // <- 2
        while let Some((cid, frame)) = frame_from_wire_receiver.next().await {
            let cid_string = cid.to_string();
            x.inc(); // <- 2 + n*2 +1
            self.metrics.streams_closed_total.with_label_values(&[&pid_string]).inc();
            trace!("handling frame");
            self.metrics.frames_in_total.with_label_values(&["dasd", "dd", "ParticipantId"]).inc();  // <- 2 + n*2 +2
            let y = self.metrics.frames_in_total.with_label_values(&["dasd", "dd", "ParticipantId"]).get();
            error!(?y, "dasdasd");  // 2 + n*2 +1
```
I would expect that &["dasd", "dd", "ParticipantId"]) is increased from 0 to 2 before the while loop.
And then 2 times for every loop. Also i would expect to trace an error and to print y which is increasing.
In fact, this also happens! Great! According to the trace the number y is increasing.
However here is the point, when accessing it over the registry, the value doesn't grow over 2!
It somehow grows till the start of the while loop, but not further.

The coding looks like
```rust
#network/examples/network-speed/metrics.rs:50
        self.handle = Some(thread::spawn(move || {
            let server = Server::new(addr, move |request| {
                router!(request,
                        (GET) (/metrics) => {
                        let encoder = TextEncoder::new();
                        let mut buffer = vec![];
                        let mf = registry.gather();
                        println!("dasd ad {:?}", mf);
                        encoder.encode(&mf, &mut buffer).expect("Failed to encoder metrics text.");
                        rouille::Response::text(String::from_utf8(buffer).expect("Failed to parse bytes as a string."))
                },
                _ => rouille::Response::empty_404()
                )
            })
                .expect("Failed to start server");
            while running2.load(Ordering::Relaxed) {
                server.poll();
                // Poll at 10Hz
                thread::sleep(Duration::from_millis(100));
            }
        }));
```

When executed and going to webbrowser 'http://127.0.0.1:59111/metrics'
I get the following return
```
# HELP channels_connected_total number of all channels currently connected on the network
# TYPE channels_connected_total counter
channels_connected_total{participant="19995010623766318553457718574023131819"} 1
# HELP connect_requests_total shows the number of connect requests to the scheduler
# TYPE connect_requests_total counter
connect_requests_total{protocol="tcp"} 1
# HELP frames_in_total number of all frames received per channel
# TYPE frames_in_total counter
frames_in_total{channel="0",frametype="Handshake",participant=""} 1
frames_in_total{channel="0",frametype="ParticipantId",participant="19995010623766318553457718574023131819"} 1
frames_in_total{channel="dd",frametype="ParticipantId",participant="dasd"} 2 // <--- HERE 2
# HELP network_info Static Network information
# TYPE network_info gauge
network_info{version="0.2.0"} 0
# HELP participants_connected_total shows the number of participants connected to the network
# TYPE participants_connected_total counter
participants_connected_total 1
# HELP participants_disconnected_total shows the number of participants disconnected to the network
# TYPE participants_disconnected_total counter
participants_disconnected_total 0
# HELP streams_opened_total number of all streams currently open on the network
# TYPE streams_opened_total counter
streams_opened_total{participant="19995010623766318553457718574023131819"} 1
```

One could think that my webserver simply doesnt update that number, but nope. i also trace the `gather` result and it shows the number 2.
Additionally after some time the stream will close and the following metrics:
```
# HELP streams_closed_total number of all streams currently open on the network
# TYPE streams_closed_total counter
streams_closed_total{participant="19995010623766318553457718574023131819"} 1
```
I also build another metrics, which will constantly increase showing that the webserver still works.
```
frames_out_total{channel="b",frametype="c",participant="a"} 716403
```
But this line wont ever increase any more
```
frames_in_total{channel="dd",frametype="ParticipantId",participant="dasd"} 2
```

If you remove one of the `x.inc()` before the while it will stay on `1`, if you add another one it will be `3`. One could probably think that the loop is never exectured... Put a `panic` inside and it will panic, also we have the tracing we see.

As you can see i am quite clueless why this behavior happens, i can only say that it consistently happen...

**To Reproduce**
1. Clone Veloren Repository and check out the prepared commit: 9e7ce3ae2e77fc14b764982ad5b34c4ab489da1f
```bash
git clone "https://gitlab.com/veloren/veloren.git"
git checkout 9e7ce3ae2e77fc14b764982ad5b34c4ab489da1f
```
2. Run the `network-speed` example coding the following way (compilation might take some time)
```bash
clear && (cd network/examples/network-speed && RUST_BACKTRACE=1 cargo run --profile=release -Z unstable-options -- --trace=trace --protocol=tcp)
```
3. while it runs, open webbrowser and open 'http://127.0.0.1:59111/metrics'. Refresh as much as you like
4. You will see  `frames_in_total{channel="dd",frametype="ParticipantId",participant="dasd"} 2` but its not increasing ever.

**Expected behavior**
```
frames_in_total{channel="dd",frametype="ParticipantId",participant="dasd"} 2
```
should increase and not stay on 2

**System information**
* CPU architecture: Ryzen 1700X - 64 bit
* Distribution and kernel version: Arch Linux, Kernel 5.6.8
* SELinux on?: No
* Any other system details we should know?: Nothing special

**Additional context**

Edit: This bug is quite heavy influincing me right now, so if i can help, e.g. by testing a fix or having a call i would be open to do so :)

**Comment by breeswish** [comment]
Thanks for the report. Currently I failed to run your program under WSL. I will try with MacOS later.

**Comment by xMAC94x** [comment]
oh it doesn't compile under windows??, weird i don't know why. do you have a specific error msg?

**Comment by breeswish** [comment]
Hi, I found that actually your `NetworkMetrics` is constructed twice, because `Scheduler::new` is called twice. This might be the root cause of your problem. 

I added the following code in `NetworkMetrics::new()` in `network/src/metrics.rs`:

```rust
        let x = frames_in_total.clone();
        std::thread::spawn(move || {
            loop {
                println!("frames_in_total collect = {:?}", x.collect());
                std::thread::sleep(std::time::Duration::from_secs(1));
            }
        });
```

It outputs two lines per second:

```
frames_in_total collect = [name: "frames_in_total" help: "number of all frames received per channel" type: COUNTER metric {label {name: "channel" value: "0"} label {name: "frametype" value: "Handshake"} label {name: "participant" value: ""} counter {value: 1}} metric {label {name: "channel" value: "0"} label {name: "frametype" value: "ParticipantId"} label {name: "participant" value: "315236423458166625005419919245515408621"} counter {value: 1}} metric {label {name: "channel" value: "dd"} label {name: "frametype" value: "ParticipantId"} label {name: "participant" value: "dasd"} counter {value: 2}}] <--------

frames_in_total collect = [name: "frames_in_total" help: "number of all frames received per channel" type: COUNTER metric {label {name: "channel" value: "0"} label {name: "frametype" value: "DataHeader"} label {name: "participant" value: "86254481468228957536731508299016678671"} counter {value: 329175}} metric {label {name: "channel" value: "0"} label {name: "frametype" value: "ParticipantId"} label {name: "participant" value: "86254481468228957536731508299016678671"} counter {value: 1}} metric {label {name: "channel" value: "0"} label {name: "frametype" value: "Handshake"} label {name: "participant" value: ""} counter {value: 1}} metric {label {name: "channel" value: "dd"} label {name: "frametype" value: "ParticipantId"} label {name: "participant" value: "dasd"} counter {value: 1316702}} <-------- .....
```

As you can see, there are actually two counter instances. One is counted correctly, and another is not counted correctly (but being registered).

I would suggest you to use a single global counter to make life easier :)

**Comment by xMAC94x** [comment]
Hi breewish. wow thanks for doing the investigation !
that is actually a quite awesome hint.

I did some research based on this, especially on the dual counter and found the source of the bug.
Luckily for you the problem was 50cm in front of my monitor ;)

In case anyone finds this on google here the problem desciption.

I Could use global version, but i want the statistics to be per `network/src/api` `Network`.
While most programs have 1 Network per binary, it will happen for the game in singleplayer mode that it will just host client and server in one binary so has 2 instances of the metrics.

The error was the following: I just monitor the client, and i code the networking speed test, that the server wont answer. thats ofc why no more messages are comming in.

I wrote proudly that the loop MUST BE executed above as i tested this with trace and panic, but the loop is only executed for the SERVER, not the client. but i am measureing the CLIENT counters because thats the one that is registered.
That explains why the server one will output all the increase and tracing etc, but the client just shows 2, as it never receives messages...

So bug can be closed, I did rookie mistake and will hopefully learn from this.

Btw, i can only say this again, thanks for your support with this "bug" and thanks for your prometheus crate. You are one of the people who keep open source growing :)
